README.md
# Timer



## Installation  

https://www.figma.com/community/plugin/820311256083321341

## Usage
Use text to create a timer on a Figma page.

There are two ways to start a timer:
1. Type the time you want to count down in **Timer: HH:MM:SS** format (e.g. Timer: 5:00), then press 'Start'.
2. Type the time you want to count down in **HH:MM:SS** format (e.g. 5:00). In this case you need to select the text layer before pressing 'Start'

Additionally, you can now  start a couple of timers on a page (for whatever reason).

## Development

1. This plugin uses Typescript & Visual Studio Code (https://code.visualstudio.com/).
To install typescript execute:

```
$ sudo npm install -g typescript
```

2. Open the directory in Visual Studio Code and perform your changes in the code.ts file
3. Run the "Terminal > Run Build Task..." menu item, then select "tsc: watch - tsconfig.json", to compile the JavaScript source.

___

Created by [Viktoriia Leontieva](https://twitter.com/killnicole), [Bernd Plontsch](https://twitter.com/berndplontsch), [Leonard Thomas](https://twitter.com/_leotho) & [Jannes Peters](https://twitter.com/jannespeters)


assets/changelog.txt
Version 3:
It is now possible to pause and reset timers.
Thanks to Jannes (🐦@jannespeters) for the contribution!

We are always happy about feedback and suggestions! https://github.com/lennet/Figma-Timer/issues/new/

code.test.ts
test('Placeholder', () => {
    expect(true).toBe(true);
  });

code.ts
const delay = ms => new Promise(res => setTimeout(res, ms));
var totalTimers = 0;
const secondsSet = [86400, 3600, 60, 1];
var pause = false;
var reset = false;
var userSetSeconds = 0;

const uiWindow = {
  minHeight: 60,
  maxHeight: 300,
  helptextHeight: 200,
  width: 220,
}

var timerUIHeight = 50;

figma.showUI(__html__, { width: uiWindow.width, height: uiWindow.minHeight })

figma.ui.onmessage = msg => {

  switch (msg.type) {

    case 'start':
      pause = false;
      reset = false;
      checkAndStart();
      break;

    case 'pause':
      pause = true;
      break;

    case 'continue':
      pause = false;
      break;

    case 'reset':
      reset = true;
      pause = true;
      totalTimers = 0;
      figma.ui.resize(uiWindow.width, uiWindow.minHeight);
      break;

    case 'helpon':
      figma.ui.resize(uiWindow.width, uiWindow.helptextHeight);
      break;

    case 'helpoff':
      figma.ui.resize(uiWindow.width, uiWindow.minHeight);
      break;

    default:
      console.log("no code for msg.type: " + msg.type);
      break;

  }
};

function checkAndStart() {
  if (checkForSelectedNodes() == false) {
    if (checkForNodesThatBeginWithTimer() == false) {
      throw new Error("Type the time to start Timer");
    }
  }
}

function checkForSelectedNodes(): boolean {
  const regex = new RegExp("[0-9]{1,2}(:[0-9]{1,2})*");
  const selectedNodes = figma.currentPage.selection.filter(node => node.type == "TEXT" && regex.test(node.characters));
  selectedNodes.forEach(start);
  return selectedNodes.length > 0;
}

function checkForNodesThatBeginWithTimer(): boolean {
  const nodes = figma.currentPage.findAll(node => node.type === "TEXT" && node.characters.startsWith("Timer:"));
  nodes.forEach(start);
  return nodes.length > 0;
}

function start(node: TextNode) {
  var timeString = node.characters;
  var startsWithTimer = false;
  if (timeString.startsWith("Timer:")) {
    timeString = timeString.replace("Timer: ", "");
    startsWithTimer = true;
  }

  var seconds = getRemainingSeconds(timeString);
  var template = getTemplateFromString(timeString);

  startTimer(node, seconds, template, startsWithTimer);

}

function getRemainingSeconds(timeString: string): number {
  var seconds = 0;
  var components = timeString.split(":");
  secondsSet.reverse();

  components.reverse().forEach((element, index) => {
    var factor = secondsSet[index];
    seconds += factor * Number(element);
  });

  secondsSet.reverse();

  return seconds;
}

/**
 * Generates a template string from a timeString
 * e.g. converts 5:00 into 0:00
 */
function getTemplateFromString(timeString: string): string {
  var result = "";
  for (const c of timeString) {
    if (c == ":") {
      result += c;
    } else {
      result += "0";
    }
  }
  return result;
}

/** 
* Creates a new time string that conforms to the templates format
* e.g. 5:00 (timeString) and 00:00:00 (template) will return 00:05:00
*/
function fillUpTimeStringWithTemplate(timeString: string, template: string): string {
  const trimmedTemplate = template.substring(0, template.length - timeString.length)
  return trimmedTemplate + timeString;
}

function secondsToInterval(seconds: number): string {
  var result = "";
  var secondsToGo = seconds;
  secondsSet.forEach((element) => {
    var count = Math.floor(secondsToGo / element);
    if (count > 0 || result.length > 0) {
      secondsToGo -= count * element;
      if (result.length > 0) {
        result += ":";
        if (count < 10) {
          result += "0";
        }
      }
      result += String(count);
    }
  })
  return result;
}


/** 
* Code that updates all timers on the Figma stage
* will also send updates / messages to UI.html, so we can show timers there
*/

async function startTimer(node: TextNode, seconds: number, template: string, startsWithTimer: boolean) {
  await figma.loadFontAsync(node.fontName as FontName);
  totalTimers += 1

  console.log("Timer started / became active");

  var timerID = totalTimers;
  var keepItRunning = true;
  var secondsToGo = seconds;
  var newText = "";

  adjustUIWindowHeight();
  figma.ui.postMessage(["start timer", newText, timerID, secondsToGo, seconds]);

  // this loop updates all timers every second
  while (keepItRunning) {

    // checking if reset was clicked by user and if so resetting all timers
    if (reset) {
      newText = fillUpTimeStringWithTemplate(secondsToInterval(seconds), template);
      //updateTimerText(startsWithTimer, newText, node);
      keepItRunning = false;
    };

    // checking if pause was NOT clicked
    if (!reset) {
      if (!pause) {
        if (secondsToGo > 0) {
          newText = fillUpTimeStringWithTemplate(secondsToInterval(secondsToGo), template);
          figma.ui.postMessage(["counting", newText, timerID, secondsToGo, seconds]);
          secondsToGo -= 1;
        } else if (secondsToGo < 1) {
          newText = "Done"
          figma.ui.postMessage(["counting", "Done", timerID, secondsToGo, seconds]);
        }
        //updateTimerText(startsWithTimer, newText, node);
      }
      updateTimerText(startsWithTimer, newText, node);
      await delay(1000);
    }
  }
  console.log("Timer finished / became in-active");
}

function updateTimerText (startsWithTimer: boolean, newText: string, node: TextNode)
{
  if (startsWithTimer) {
    newText = "Timer: " + newText;
  }
  node.characters = newText;
}

// adjusting height of UI windows depending on amount of timers
function adjustUIWindowHeight() {
  var newUIHeight = 100 + totalTimers * 50;
  if (newUIHeight > uiWindow.maxHeight) {
    newUIHeight = uiWindow.maxHeight;
  }
  figma.ui.resize(uiWindow.width, newUIHeight);
}

code.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
const delay = ms => new Promise(res => setTimeout(res, ms));
var totalTimers = 0;
const secondsSet = [86400, 3600, 60, 1];
var pause = false;
var reset = false;
var userSetSeconds = 0;
const uiWindow = {
    minHeight: 60,
    maxHeight: 300,
    helptextHeight: 200,
    width: 220,
};
var timerUIHeight = 50;
figma.showUI(__html__, { width: uiWindow.width, height: uiWindow.minHeight });
figma.ui.onmessage = msg => {
    switch (msg.type) {
        case 'start':
            pause = false;
            reset = false;
            checkAndStart();
            break;
        case 'pause':
            pause = true;
            break;
        case 'continue':
            pause = false;
            break;
        case 'reset':
            reset = true;
            pause = true;
            totalTimers = 0;
            figma.ui.resize(uiWindow.width, uiWindow.minHeight);
            break;
        case 'helpon':
            figma.ui.resize(uiWindow.width, uiWindow.helptextHeight);
            break;
        case 'helpoff':
            figma.ui.resize(uiWindow.width, uiWindow.minHeight);
            break;
        default:
            console.log("no code for msg.type: " + msg.type);
            break;
    }
};
function checkAndStart() {
    if (checkForSelectedNodes() == false) {
        if (checkForNodesThatBeginWithTimer() == false) {
            throw new Error("Type the time to start Timer");
        }
    }
}
function checkForSelectedNodes() {
    const regex = new RegExp("[0-9]{1,2}(:[0-9]{1,2})*");
    const selectedNodes = figma.currentPage.selection.filter(node => node.type == "TEXT" && regex.test(node.characters));
    selectedNodes.forEach(start);
    return selectedNodes.length > 0;
}
function checkForNodesThatBeginWithTimer() {
    const nodes = figma.currentPage.findAll(node => node.type === "TEXT" && node.characters.startsWith("Timer:"));
    nodes.forEach(start);
    return nodes.length > 0;
}
function start(node) {
    var timeString = node.characters;
    var startsWithTimer = false;
    if (timeString.startsWith("Timer:")) {
        timeString = timeString.replace("Timer: ", "");
        startsWithTimer = true;
    }
    var seconds = getRemainingSeconds(timeString);
    var template = getTemplateFromString(timeString);
    startTimer(node, seconds, template, startsWithTimer);
}
function getRemainingSeconds(timeString) {
    var seconds = 0;
    var components = timeString.split(":");
    secondsSet.reverse();
    components.reverse().forEach((element, index) => {
        var factor = secondsSet[index];
        seconds += factor * Number(element);
    });
    secondsSet.reverse();
    return seconds;
}
/**
 * Generates a template string from a timeString
 * e.g. converts 5:00 into 0:00
 */
function getTemplateFromString(timeString) {
    var result = "";
    for (const c of timeString) {
        if (c == ":") {
            result += c;
        }
        else {
            result += "0";
        }
    }
    return result;
}
/**
* Creates a new time string that conforms to the templates format
* e.g. 5:00 (timeString) and 00:00:00 (template) will return 00:05:00
*/
function fillUpTimeStringWithTemplate(timeString, template) {
    const trimmedTemplate = template.substring(0, template.length - timeString.length);
    return trimmedTemplate + timeString;
}
function secondsToInterval(seconds) {
    var result = "";
    var secondsToGo = seconds;
    secondsSet.forEach((element) => {
        var count = Math.floor(secondsToGo / element);
        if (count > 0 || result.length > 0) {
            secondsToGo -= count * element;
            if (result.length > 0) {
                result += ":";
                if (count < 10) {
                    result += "0";
                }
            }
            result += String(count);
        }
    });
    return result;
}
/**
* Code that updates all timers on the Figma stage
* will also send updates / messages to UI.html, so we can show timers there
*/
function startTimer(node, seconds, template, startsWithTimer) {
    return __awaiter(this, void 0, void 0, function* () {
        yield figma.loadFontAsync(node.fontName);
        totalTimers += 1;
        console.log("Timer started / became active");
        var timerID = totalTimers;
        var keepItRunning = true;
        var secondsToGo = seconds;
        var eventType = "start timer";
        var newText = "";
        adjustUIWindowHeight();
        postMessageToUIWindow(eventType, newText, timerID, secondsToGo, seconds);
        // this loop updates all timers every second
        while (keepItRunning) {
            // checking if reset was clicked by user and if so resetting all timers
            if (reset) {
                newText = fillUpTimeStringWithTemplate(secondsToInterval(seconds), template);
                keepItRunning = false;
                updateTimerText(startsWithTimer, newText, node);
            }
            else if (!pause) {
                if (secondsToGo > 0) {
                    newText = fillUpTimeStringWithTemplate(secondsToInterval(secondsToGo), template);
                    eventType = "counting";
                }
                else {
                    newText = "Done";
                    eventType = "timer done";
                }
                postMessageToUIWindow(eventType, newText, timerID, secondsToGo, seconds);
                updateTimerText(startsWithTimer, newText, node);
                secondsToGo -= 1;
            }
            yield delay(1000);
        }
        console.log("Timer finished / became in-active");
    });
}
function postMessageToUIWindow(eventType, timerText, timerID, secondsToGo, secondsToStart) {
    figma.ui.postMessage([eventType, timerText, timerID, secondsToGo, secondsToStart]);
}
function updateTimerText(startsWithTimer, newText, node) {
    if (startsWithTimer) {
        newText = "Timer: " + newText;
    }
    node.characters = newText;
}
// adjusting height of UI windows depending on amount of timers
function adjustUIWindowHeight() {
    var newUIHeight = 100 + totalTimers * 50;
    if (newUIHeight > uiWindow.maxHeight) {
        newUIHeight = uiWindow.maxHeight;
    }
    figma.ui.resize(uiWindow.width, newUIHeight);
}
